### PR TITLE
RDS: Add DBSubnet resource, VPC security groups

### LINKED ISSUE #6: When creating a RDS in a VPC, the RDS Security Group cannot be created
First of all, thank you for this project, it has been great so far.

When I try to create an RDS Security Group and an RDS instance, nixops deploy exits and fails with:

```botocore.exceptions.ClientError: An error occurred (InvalidParameterValue) when calling the CreateDBSecurityGroup operation: The vpc none does not exist.```

Though I only specified the SecurityGroupId in my rules (using the key-format specified in the code) because all of my stuff lives in VPCs.

Any idea?

### PR DESCRIPTION
**Pull request by RaitoBezarius**
This commit enables EC2 RDS resource to use:

- DBSubnets
- VPC security groups

As a result, it adds a DBSubnet resource which consists of at least 2
VPC subnets containing at least each AZ of your region.

It enables support of RDS in a VPC.

This makes it possible to use RDS in a VPC setup, i.e. in EC2-VPC.

Will fix #6.

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
It has been 300 minutes but it's done :) @AmineChikhaoui

**Comment by AmineChikhaoui** [comment]
No worries @RaitoBezarius , thanks !
I'll try to play with it later this evening.

File: nix/ec2-rds-subnet-group.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    region = mkOption {
+      type = types.str;
+      description = "Amazon RDS DB subnet group region.";
+    };
+
+    accessKeyId = mkOption {
+      default = "";
+      type = types.str;
+      description = "The AWS Access Key ID.";
+    };
```
Comment (PsyanticY) [inline nix/ec2-rds-subnet-group.nix]: u can use `imports = [ ./common-ec2-auth-options.nix ];` for these options like we do in the other modules [for example](https://github.com/NixOS/nixops-aws/blob/master/nix/vpc-endpoint.nix)

File: nix/ec2-rds-subnet-group.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    subnetIds = mkOption {
+      default = [];
+      type = types.listOf (types.either types.str (resource "vpc-subnet"));
+      apply = map (x: if builtins.isString x then x else "res-" + x._name + "." + x._type + ".subnetId");
+      description = "List of VPC subnets to use for this DB subnet group (must contain at least a subnet for each AZ), it can be a VPC Subnet resource or a string representing the ID of the VPC Subnet.";
```
Comment (PsyanticY) [inline nix/ec2-rds-subnet-group.nix]: break into multiple lines.

**Comment by PsyanticY** [comment]
Update this [line](https://github.com/NixOS/nixops-aws/pull/13/files#diff-42cf3b4dfb22cdbc05c9132d9db8cadfR192) to `if dbinstance.status not in {'creating', 'backing-up', 'available', 'modifying', 'configuring-enhanced-monitoring'}:`

File: examples/ec2-rds-with-vpc.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        id = "test-multi-az";
+        instanceClass = "db.r3.large";
+        allocatedStorage = 30;
+        masterUsername = "administrator";
+        masterPassword = "testing123";
+        port = 5432;
+        engine = "postgres";
+        dbName = "testNixOps";
+        multiAZ = true;
+        vpcSecurityGroups = [ resources.ec2SecurityGroups.database ];
```
Comment (PsyanticY) [inline examples/ec2-rds-with-vpc.nix]: should be `vpcSecurityGroupIds`

File: nix/ec2-rds-subnet-group.nix (hunk 1 above)
Comment (RaitoBezarius) [reply nix/ec2-rds-subnet-group.nix]: Sure.

File: nix/ec2-rds-subnet-group.nix (hunk 2 above)
Comment (RaitoBezarius) [reply nix/ec2-rds-subnet-group.nix]: Okay!

File: examples/ec2-rds-with-vpc.nix (hunk 3 above)
Comment (RaitoBezarius) [reply examples/ec2-rds-with-vpc.nix]: Right!

**Comment by RaitoBezarius** [comment]
> Update this [line](https://github.com/NixOS/nixops-aws/pull/13/files#diff-42cf3b4dfb22cdbc05c9132d9db8cadfR192) to `if dbinstance.status not in {'creating', 'backing-up', 'available', 'modifying', 'configuring-enhanced-monitoring'}:`

Why this change?

File: nix/ec2-rds-dbinstance.nix [hunk 4]
```diff
@@ -86,10 +86,28 @@ with import ./lib.nix lib;
...
+        List of names of VPCSecurityGroupMembership to authorize on this DBInstance, use this if you are in an VPC and not on EC2-Classic. Not applicable for Amazon Aurora.
+        '';
+    };
+
+    dbSubnetGroup = mkOption {
+      default = null;
+      type = types.nullOr (types.either types.str (resource "ec2-rds-db-subnet-group"));
+      apply = group: if (builtins.isString group || group == null) then group else "res-" + group._name;
+      description = ''
+        A name for a DBSubnetGroup, they must contain at least one subnet in each availability zone in the AWS region.
```
Comment (PsyanticY) [inline nix/ec2-rds-dbinstance.nix]: multi lines

**Comment by PsyanticY** [comment]
Tested deploying the example after some minor changes. and it worked fine. We have an error in destroying the example since nixops do not wait for the rds to finalize termination and fail when destroying the security group cause it is still attached to the `delete in progress` rds.
```
[nix-shell:~/rds/nixops]$ nixops destroy -d rds-test
warning: are you sure you want to destroy RDS instance ‘test-multi-az’? (y/N) y
test-rds-instance> deleting RDS instance `test-multi-az'...
test-rds-instance> saving final snapshot as test-multi-az-final-snapshot-535c8be39f404c32aa949bb5de2ea1e1
db-subnet........> destroying rds db subnet group nixops-c6b58b14-fcba-11e9-b33e-0ae168952ac0-db-subnet
database.........> deleting EC2 security group `charon-c6b58b14-fcba-11e9-b33e-0ae168952ac0-database' ID `sg-05e1714ce3a1603e7'...
error: Multiple exceptions (2): 
  * database: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>DependencyViolation</Code><Message>resource sg-05e1714ce3a1603e7 has a dependent object</Message></Error></Errors><RequestID>1f56d3bf-0b36-4064-bf5a-eb5215bcc515</RequestID></Response>
  * db-subnet: An error occurred (InvalidDBSubnetGroupStateFault) when calling the DeleteDBSubnetGroup operation: Cannot delete the subnet group 'nixops-c6b58b14-fcba-11e9-b33e-0ae168952ac0-db-subnet' because at least one database instance: test-multi-az is still using it.
```

We should adress this. also I think we need to add the backup functionality to RDS module as well as remove automatically taking final snapshot (make a prompt or an argument).
@AmineChikhaoui

**Comment by RaitoBezarius** [comment]
> Tested deploying the example after some minor changes. and it worked fine. We have an error in destroying the example since nixops do not wait for the rds to finalize termination and fail when destroying the security group cause it is still attached to the `delete in progress` rds.
> 
> ```
> [nix-shell:~/rds/nixops]$ nixops destroy -d rds-test
> warning: are you sure you want to destroy RDS instance ‘test-multi-az’? (y/N) y
> test-rds-instance> deleting RDS instance `test-multi-az'...
> test-rds-instance> saving final snapshot as test-multi-az-final-snapshot-535c8be39f404c32aa949bb5de2ea1e1
> db-subnet........> destroying rds db subnet group nixops-c6b58b14-fcba-11e9-b33e-0ae168952ac0-db-subnet
> database.........> deleting EC2 security group `charon-c6b58b14-fcba-11e9-b33e-0ae168952ac0-database' ID `sg-05e1714ce3a1603e7'...
> error: Multiple exceptions (2): 
>   * database: EC2ResponseError: 400 Bad Request
> <?xml version="1.0" encoding="UTF-8"?>
> <Response><Errors><Error><Code>DependencyViolation</Code><Message>resource sg-05e1714ce3a1603e7 has a dependent object</Message></Error></Errors><RequestID>1f56d3bf-0b36-4064-bf5a-eb5215bcc515</RequestID></Response>
>   * db-subnet: An error occurred (InvalidDBSubnetGroupStateFault) when calling the DeleteDBSubnetGroup operation: Cannot delete the subnet group 'nixops-c6b58b14-fcba-11e9-b33e-0ae168952ac0-db-subnet' because at least one database instance: test-multi-az is still using it.
> ```
> 
> We should adress this. also I think we need to add the backup functionality to RDS module as well as remove automatically taking final snapshot (make a prompt or an argument).
> @AmineChikhaoui

FYI, I noticed this issue and opened #11 for it, +1 for the final snapshot, I think it's a sane default to have though.

File: nix/ec2-rds-subnet-group.nix [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    };
+
+    description = mkOption {
+      type = types.str;
+      description = ''
+        Description of the RDS DB subnet group.
+      ''; 
+    };
+
+    imports = [ ./common-ec2-options.nix ];
```
Comment (AmineChikhaoui) [inline nix/ec2-rds-subnet-group.nix]: This addition wasn't tested, can you please make sure that it works ?
It doesn't evaluate as ./common-ec2-options.nix is a function so please make it's merged to options
```
options = {
   /* options declarations */
} // (import ./common-ec2-options.nix { inherit lib; });
```
Also if I fix that locally the diff engine will throw an error as it doesn't find a handler that realizes the `tags` diff, so it needs to be fixed in the python side to create a handler for the `tags` option.

File: examples/ec2-rds-with-vpc.nix [hunk 6]
```diff
@@ -0,0 +1,71 @@
+let
+  region = "us-east-1";
+  accessKeyId = "AKIA...";
+in
+  {
+    network.description = "NixOps RDS in a VPC Testing";
+  # A VPC.
+  resources.vpc.private = {
```
Comment (AmineChikhaoui) [inline examples/ec2-rds-with-vpc.nix]: Can you also fix the indentation in this file please and remove any comments such as `# A VPC` which don't seem to be useful.

File: nixopsaws/resources/ec2_rds_dbinstance.py [hunk 7]
```diff
@@ -91,7 +104,14 @@ def resource_id(self):
 
     def create_after(self, resources, defn):
         return {r for r in resources if
-                isinstance(r, nixopsaws.resources.ec2_rds_dbsecurity_group.EC2RDSDbSecurityGroupState)}
+                isinstance(r, nixopsaws.resources.ec2_rds_dbsecurity_group.EC2RDSDbSecurityGroupState)
+                or isinstance(r, nixopsaws.resources.ec2_rds_subnet_group.EC2RDSSubnetGroupState)
+                or isinstance(r, nixopsaws.resources.ec2_security_group.EC2SecurityGroupState)}
+
+    def destroy_before(self, resources):
```
Comment (AmineChikhaoui) [inline nixopsaws/resources/ec2_rds_dbinstance.py]: Do we really need `destroy_before` ? afaik it's automatically handled by doing the opposite of create_after

Comment (RaitoBezarius) [reply nixopsaws/resources/ec2_rds_dbinstance.py]: I didn't know this, as I was exploring the solution for #11

**Comment by RaitoBezarius** [comment]
@AmineChikhaoui Changes were made.

**Comment by AmineChikhaoui** [comment]
Thank you for the quick changes but *please* test before asking for review again :)

**Comment by RaitoBezarius** [comment]
@AmineChikhaoui Now it is tested.

**Comment by AmineChikhaoui** [comment]
@RaitoBezarius I think there might be some minor changes to be done to make sure that when not using VPC security groups/DB subnet the config works as well.
I get this with the old example of RDS:
```
[nix-shell:~/src/nixops]$ nixops deploy -d rds-legacy
Traceback (most recent call last):
  File "/home/amine/src/nixops/scripts/nixops", line 250, in <module>
    args.op(args)
  File "/home/amine/src/nixops/nixops/script_defs.py", line 427, in op_deploy
    max_concurrent_activate=args.max_concurrent_activate)
  File "/home/amine/src/nixops/nixops/deployment.py", line 1062, in deploy
    self.run_with_notify('deploy', lambda: self._deploy(**kwargs))
  File "/home/amine/src/nixops/nixops/deployment.py", line 1051, in run_with_notify
    f()
  File "/home/amine/src/nixops/nixops/deployment.py", line 1062, in <lambda>
    self.run_with_notify('deploy', lambda: self._deploy(**kwargs))
  File "/home/amine/src/nixops/nixops/deployment.py", line 915, in _deploy
    self.evaluate_active(include, exclude, kill_obsolete)
  File "/home/amine/src/nixops/nixops/deployment.py", line 877, in evaluate_active
    self.evaluate()
  File "/home/amine/src/nixops/nixops/deployment.py", line 383, in evaluate
    defn = _create_definition(y, config["resources"][res_type][name], res_type)
  File "/home/amine/src/nixops/nixops/deployment.py", line 1265, in _create_definition
    return cls(xml)
  File "/nix/store/nihlv8ac909rw3lca1jdhalx3j72sw1x-nixops-aws-1.7pre0_abcdef/lib/python2.7/site-packages/nixopsaws/resources/ec2_rds_dbinstance.py", line 49, in __init__
    self.rds_dbinstance_db_subnet_group = xml.find("attrs/attr[@name='dbSubnetGroup']/string").get("value")
AttributeError: 'NoneType' object has no attribute 'get'
```

**Comment by RaitoBezarius** [comment]
@AmineChikhaoui I made the change but I'm unable to test it, I would require an ol'EC2 Classic account, those are not available anymore as AWS defaults to EC2-VPC nowadays AFAIK.

I tried to run it but AWS throws me an error saying: `botocore.exceptions.ClientError: An error occurred (InvalidParameterValue) when calling the AuthorizeDBSecurityGroupIngress operation: VPC DB Security Groups cannot be modified with this API version.  Please use an API version between 2012-01-15 and 2012-10-31 to modify this group.`

But according to this: https://github.com/hashicorp/terraform/issues/486 it's expected for people having an EC2 VPC account (like myself).

Let me know your thoughts on this.

**Comment by NickSeagull** [comment]
Any progress on this?

**Comment by RaitoBezarius** [comment]
@NickSeagull AFAIK, I'm waiting for @AmineChikhaoui to give it a try again and see what should we do regarding the EC2-Classic stuff that I cannot test.

**Comment by AmineChikhaoui** [comment]
Last time I tested on ec2-classic there was still some problems, I'll try to post the stacktrace next time but I'm not sure how helpful that would be unless I'm going to do the fixes as you can't test it.

**Comment by RaitoBezarius** [comment]
@AmineChikhaoui It might help us to move forward, I wonder whether I can get my hand on an EC2 Classic account, I'll try to ask AWS support staff.

**Comment by RaitoBezarius** [comment]
Closing because of #113

**Pull request closed without merging**